File: scripts/check_gherkin_format.py
#!/usr/bin/env python3
"""人が読める日本語Gherkinテスト起票を安全側で検査する。"""

from __future__ import annotations

import re
import sys
from dataclasses import dataclass, field
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
FEATURE_ROOT = ROOT / "test" / "features"
LAYERS = {"unit", "integration", "e2e"}
JAPANESE = re.compile(r"[\u3040-\u30ff\u3400-\u9fff]")
HEADER = re.compile(r"^\s*(Feature|Scenario(?: Outline)?|Given|When|Then|And|But):?\s+(.+?)\s*$")
SCENARIO = re.compile(r"^\s*Scenario(?: Outline)?:\s+(SCN-[A-Z0-9-]+)\s+(.+?)\s*$")
STEP = re.compile(r"^\s*(Given|When|Then|And|But)\s+(.+?)\s*$")


@dataclass
class ScenarioRecord:
    scenario_id: str
    file: Path
    line: int
    keywords: set[str] = field(default_factory=set)


def relative(path: Path) -> str:
    return path.relative_to(ROOT).as_posix()


def check() -> list[str]:
    errors: list[str] = []
    files = sorted(FEATURE_ROOT.glob("**/*.feature"))
    if not files:
        return ["test/features配下に.featureがありません"]

    scenarios: list[ScenarioRecord] = []
    layer_counts = {layer: 0 for layer in LAYERS}
    for file in files:
        parts = set(file.relative_to(FEATURE_ROOT).parts)
        layers = parts & LAYERS
        if len(layers) != 1:
            errors.append(f"{relative(file)}: unit/integration/e2eのいずれか1 layer配下に置いてください")
        else:
            layer_counts[next(iter(layers))] += 1

        current: ScenarioRecord | None = None
        feature_seen = False
        for number, line in enumerate(file.read_text(encoding="utf-8").splitlines(), start=1):
            stripped = line.strip()
            if not stripped or stripped.startswith("#") or stripped.startswith("@") or stripped.startswith("|"):
                continue
            if stripped.startswith("Feature:"):
                feature_seen = True
            match = HEADER.match(line)
            if match and not JAPANESE.search(match.group(2)):
                errors.append(f"{relative(file)}:{number}: {match.group(1)}の説明本文を日本語で記述してください")
            scenario = SCENARIO.match(line)
            if scenario:
                current = ScenarioRecord(scenario.group(1), file, number)
                scenarios.append(current)
                continue
            step = STEP.match(line)
            if step and current:
                current.keywords.add(step.group(1))
        if not feature_seen:
            errors.append(f"{relative(file)}: Featureがありません")

    seen: dict[str, ScenarioRecord] = {}
    for scenario in scenarios:
        if scenario.scenario_id in seen:
            first = seen[scenario.scenario_id]
            errors.append(
                f"{relative(scenario.file)}:{scenario.line}: SCN ID {scenario.scenario_id}が重複しています "
                f"(最初の定義: {relative(first.file)}:{first.line})"
            )
        else:
            seen[scenario.scenario_id] = scenario
        for keyword in ("Given", "When", "Then"):
            if keyword not in scenario.keywords:
                errors.append(f"{relative(scenario.file)}:{scenario.line}: {scenario.scenario_id}に{keyword}がありません")

    for layer, count in sorted(layer_counts.items()):
        if count == 0:
            errors.append(f"{layer} layerに.featureがありません")

    node_tests = sorted((ROOT / "test").glob("**/*.test.js"))
    for file in node_tests:
        errors.append(f"{relative(file)}: Node test起票を残さずGherkin .featureへ移行してください")

    return errors


def main() -> int:
    errors = check()
    if errors:
        print("Gherkin形式検査: 失敗", file=sys.stderr)
        for error in errors:
            print(f"- {error}", file=sys.stderr)
        return 1
    print("Gherkin形式検査: 合格（英語keyword・日本語説明、一意のSCN ID、全テスト層）")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/check_gherkin_format.ts
import fs from "node:fs";
import path from "node:path";

const root = path.resolve(import.meta.dirname, "..");
const featureRoot = path.join(root, "test", "features");
const layers = new Set(["unit", "integration", "e2e"]);
const japanese = /[\u3040-\u30ff\u3400-\u9fff]/u;
const header =
  /^\s*(Feature|Scenario(?: Outline)?|Given|When|Then|And|But):?\s+(.+?)\s*$/u;
const scenarioPattern =
  /^\s*Scenario(?: Outline)?:\s+(SCN-[A-Z0-9-]+)\s+(.+?)\s*$/u;
const step = /^\s*(Given|When|Then|And|But)\s+(.+?)\s*$/u;

interface ScenarioRecord {
  id: string;
  file: string;
  line: number;
  keywords: Set<string>;
}

function walk(directory: string, suffix: string): string[] {
  if (!fs.existsSync(directory)) return [];
  return fs.readdirSync(directory, { withFileTypes: true }).flatMap((entry) => {
    const current = path.join(directory, entry.name);
    if (entry.isDirectory()) return walk(current, suffix);
    return entry.isFile() && entry.name.endsWith(suffix) ? [current] : [];
  });
}

const relative = (file: string): string =>
  path.relative(root, file).replaceAll(path.sep, "/");

export function checkGherkinFormat(): string[] {
  const files = walk(featureRoot, ".feature").sort();
  if (files.length === 0) return ["test/features配下に.featureがありません"];
  const errors: string[] = [];
  const scenarios: ScenarioRecord[] = [];
  const layerCounts = new Map([...layers].map((layer) => [layer, 0]));
  for (const file of files) {
    const selectedLayers = path
      .relative(featureRoot, file)
      .split(path.sep)
      .filter((part) => layers.has(part));
    if (selectedLayers.length !== 1)
      errors.push(
        `${relative(file)}: unit/integration/e2eのいずれか1 layer配下に置いてください`,
      );
    else
      layerCounts.set(
        selectedLayers[0]!,
        (layerCounts.get(selectedLayers[0]!) ?? 0) + 1,
      );
    let current: ScenarioRecord | undefined;
    let featureSeen = false;
    const lines = fs.readFileSync(file, "utf8").split(/\r?\n/u);
    for (const [index, line] of lines.entries()) {
      const stripped = line.trim();
      if (stripped.length === 0 || /^[#@|]/u.test(stripped)) continue;
      if (stripped.startsWith("Feature:")) featureSeen = true;
      const headerMatch = header.exec(line);
      if (headerMatch && !japanese.test(headerMatch[2]!))
        errors.push(
          `${relative(file)}:${index + 1}: ${headerMatch[1]}の説明本文を日本語で記述してください`,
        );
      const scenarioMatch = scenarioPattern.exec(line);
      if (scenarioMatch) {
        current = {
          id: scenarioMatch[1]!,
          file,
          line: index + 1,
          keywords: new Set(),
        };
        scenarios.push(current);
        continue;
      }
      const stepMatch = step.exec(line);
      if (stepMatch && current) current.keywords.add(stepMatch[1]!);
    }
    if (!featureSeen) errors.push(`${relative(file)}: Featureがありません`);
  }
  const seen = new Map<string, ScenarioRecord>();
  for (const scenario of scenarios) {
    const first = seen.get(scenario.id);
    if (first)
      errors.push(
        `${relative(scenario.file)}:${scenario.line}: SCN ID ${scenario.id}が重複しています (最初の定義: ${relative(first.file)}:${first.line})`,
      );
    else seen.set(scenario.id, scenario);
    for (const keyword of ["Given", "When", "Then"])
      if (!scenario.keywords.has(keyword))
        errors.push(
          `${relative(scenario.file)}:${scenario.line}: ${scenario.id}に${keyword}がありません`,
        );
  }
  for (const [layer, count] of [...layerCounts].sort())
    if (count === 0) errors.push(`${layer} layerに.featureがありません`);
  for (const file of walk(path.join(root, "test"), ".test.js"))
    errors.push(
      `${relative(file)}: Node test起票を残さずGherkin .featureへ移行してください`,
    );
  return errors;
}

const errors = checkGherkinFormat();
if (errors.length > 0) {
  process.stderr.write(
    `Gherkin形式検査: 失敗\n${errors.map((error) => `- ${error}`).join("\n")}\n`,
  );
  process.exitCode = 1;
} else
  process.stdout.write(
    "Gherkin形式検査: 合格（英語keyword・日本語説明、一意のSCN ID、全テスト層）\n",
  );
